File: 16_02_2024_Least_Number_of_Unique_Integers_after_K_Removals.cpp
class Solution {
public:
    int findLeastNumOfUniqueInts(vector<int>& arr, int k) {
        unordered_map<int,int> mpp;
        multimap<int,int>npp;
        for(int i : arr)
        {
            mpp[i]++;
        }
        for(auto it : mpp)
        {
            npp.insert({it.second,it.first});
        }
        for(auto it:npp)
        {
            if(k<=0)
            break;
            else
            {
                k-=it.first;
                mpp.erase(mpp.find(it.second));
            }
        }
        if(k<0)
        return mpp.size()+1;
        return mpp.size();
    }
};

File: 16_02_2024_Least_Number_of_Unique_Integers_after_K_Removals.java
public class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        HashMap<Integer, Integer> mpp = new HashMap<>();
        PriorityQueue<Map.Entry<Integer, Integer>> npp = new PriorityQueue<>(
                (a, b) -> a.getValue() - b.getValue()
        );
        for (int i : arr) {
            mpp.put(i, mpp.getOrDefault(i, 0) + 1);
        }
        npp.addAll(mpp.entrySet());
        while (k > 0 && !npp.isEmpty()) {
            Map.Entry<Integer, Integer> entry = npp.poll();
            if (k >= entry.getValue()) {
                k -= entry.getValue();
                mpp.remove(entry.getKey());
            } else {
                break;
            }
        }
        return mpp.size();
    }
}

